# 수화 이미지 분류 경진대회 Baseline 2

안녕하세요 데이커 여러분! 이번 대회는 주어진 수화 이미지의 숫자를 분류하는 대회입니다.

이전 베이스라인은 잘 보셨나요?

이번 베이스라인에서는 딥러닝 모델 구축 시 많이 사용되는 기법인 전이학습(transfer learning) 에 대해 소개하고자 합니다.

전이학습이란 이미 구축되어 있는 모델을 사용하는 기법입니다.

그럼 본격적으로 코드를 통해 전이학습에 대해 알아보아요!

베이스라인을 통해 이미지 처리 기초에 입문해보세요!

<!-- 이번 베이스라인 코드에서는 pytorch를 활용해 단순 CNN 모델을 사용하여 이미지 분류를 해보았습니다.
CNN 을 활용한 알고리즘은 컴퓨터 비전 분야의 기초가 되는 알고리즘입니다! 
그럼 코드와 설명을 보고 CNN 알고리즘을 이해해 봅시다! 

* 코드를 어떻게 실행시켜야 할지 잘 모르시는 분은 아래 "코랩으로 데이콘 참여하기"를 먼저 봐주세요!
https://dacon.io/competitions/official/235836/talkboard/404882

* 데이터를 살펴보는 탐색적 데이터 분석 (Exploratory Data Analysis, EDA) 코드를 먼저 보고 오시면 좋습니다. -->


## 환경 설정 및 데이터 로드
GPU를 쓸수 있는 환경이면 GPU 부터 할당해보도록 하겠습니다.

In [1]:
import os 
import torch
device = torch.device('cuda') if torch.cuda.is_available() else torch.device('cpu') #GPU 할당

하이퍼 파라미터 값을 지정하겠습니다.    

In [2]:
#하이퍼 파라미터 튜닝

CFG = {
    'IMG_SIZE':128, #이미지 사이즈
    'EPOCHS':50, #에포크
    'LEARNING_RATE':2e-2, #학습률
    'BATCH_SIZE':12, #배치사이즈
    'SEED':41, #시드
}

모델의 재현성을 위하여 random seed를 고정하겠습니다.

In [3]:
# Seed 고정
import random
import numpy as np

def seed_everything(seed):
    random.seed(seed)
    os.environ['PYTHONHASHSEED'] = str(seed)
    np.random.seed(seed)
    torch.manual_seed(seed)
    torch.cuda.manual_seed(seed)
    torch.backends.cudnn.deterministic = True
    torch.backends.cudnn.benchmark = True

seed_everything(CFG['SEED'])

먼저 csv 파일을 불러와서 label이 어떻게 되어있는지 살펴보겠습니다.

In [4]:
import pandas as pd
label_df = pd.read_csv('data/train.csv')
label_df.head()

file_name label
0   001.png  10-2
1   002.png  10-1
2   003.png     3
3   004.png     8
4   005.png     9

In [5]:
label_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 858 entries, 0 to 857
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   file_name  858 non-null    object
 1   label      858 non-null    object
dtypes: object(2)
memory usage: 13.5+ KB


label이 object 타입으로 되어있는 것을 확인할 수 있습니다.   
또한 10을 나타내는 수화 사진이 10-1, 10-2로 두가지로 분류되어있는 것을 알 수 있습니다.   

## 데이터 전처리

모델링을 하기 위해 label을 정수형으로 수정해 주겠습니다.   
파이토치는 클래스의 인덱스 번호가 0부터 읽어들이기 때문에,   
10-1를 정수 10으로, 10를 정수 0으로 바꿔주고   
label 열의 타입을 int로 수정해 주겠습니다.

In [6]:
label_df['label'][label_df['label'] == '10-1'] = 10 ## label : 10-1 -> 10
label_df['label'][label_df['label'] == '10-2'] = 0 ## Label : 10-2 -> 0
label_df['label'] = label_df['label'].apply(lambda x : int(x)) ## Dtype : object -> int

In [7]:
label_df.head()

file_name  label
0   001.png      0
1   002.png     10
2   003.png      3
3   004.png      8
4   005.png      9

In [8]:
label_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 858 entries, 0 to 857
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   file_name  858 non-null    object
 1   label      858 non-null    int64 
dtypes: int64(1), object(1)
memory usage: 13.5+ KB


head()에서 label 열과 info()에서 Dtype을 보면    
정상적으로 수정된 것을 확인 할 수 있습니다.

그리고 데이터 이미지의 local adress와 label 값을 list에 저장해주도록 하겠습니다.

In [9]:
from glob import glob

def get_train_data(data_dir):
    img_path_list = []
    label_list = []
    
    # get image path
    img_path_list.extend(glob(os.path.join(data_dir, '*.png')))
    img_path_list.sort(key=lambda x:int(x.split('/')[-1].split('.')[0]))
    #print('wow', img_path_list)
        
    # get label
    #label_df = pd.read_csv(data_dir+'/train.csv')
    label_list.extend(label_df['label'])
    #print('wow2', label_list)
                
    return img_path_list, label_list

def get_test_data(data_dir):
    img_path_list = []
    
    # get image path
    img_path_list.extend(glob(os.path.join(data_dir, '*.png')))
    img_path_list.sort(key=lambda x:int(x.split('/')[-1].split('.')[0]))
    #print(img_path_list)
    
    return img_path_list

In [10]:
all_img_path, all_label = get_train_data('data/train')
test_img_path = get_test_data('data/test')

label값 상위 5개를 확인해봅시다.

In [12]:
all_label[0:5]

[0, 10, 3, 8, 9]

## CustomDataset

전체 dataset을 구성하는 단계입니다.  

In [13]:
import torchvision.datasets as datasets # 이미지 데이터셋 집합체
import torchvision.transforms as transforms # 이미지 변환 툴

from torch.utils.data import DataLoader # 학습 및 배치로 모델에 넣어주기 위한 툴
from torch.utils.data import DataLoader, Dataset

class CustomDataset(Dataset):
    def __init__(self, img_path_list, label_list, train_mode=True, transforms=None): #필요한 변수들을 선언
        self.transforms = transforms
        self.train_mode = train_mode
        self.img_path_list = img_path_list
        self.label_list = label_list

    def __getitem__(self, index): #index번째 data를 return
        img_path = self.img_path_list[index]
        # Get image data
        image = cv2.imread(img_path)
        if self.transforms is not None:
            image = self.transforms(image)

        if self.train_mode:
            label = self.label_list[index]
            return image, label
        else:
            return image
    
    def __len__(self): #길이 return
        return len(self.img_path_list)

#### Train / Validation Split

그럼 학습시킬 데이터 셋과 검증할 데이터 셋을 분리해주도록 하겠습니다.

In [14]:
# Train : Validation = 0.8 : 0.25 Split
train_len = int(len(all_img_path)*0.75)
Vali_len = int(len(all_img_path)*0.25)

train_img_path = all_img_path[:train_len]
train_label = all_label[:train_len]

vali_img_path = all_img_path[train_len:]
vali_label = all_label[train_len:]

In [15]:
print('train set 길이 : ', train_len)
print('vaildation set 길이 : ', Vali_len)

train set 길이 :  643
vaildation set 길이 :  214


train set은 643개, vaildation set은 214개로 나뉘어진 것을 확인할 수 있습니다.  
그럼 나뉜 데이터 셋에서 이미지를 분석 하기 위해 이미지 변형(transform)을 적용해보도록 하겠습니다.

In [16]:
train_transform = transforms.Compose([
                    transforms.ToPILImage(), #Numpy배열에서 PIL이미지로
                    transforms.Resize([CFG['IMG_SIZE'], CFG['IMG_SIZE']]), #이미지 사이즈 변형
                    transforms.ToTensor(), #이미지 데이터를 tensor
                    transforms.Normalize(mean=(0.5, 0.5, 0.5), std=(0.5, 0.5, 0.5)) #이미지 정규화
                    
                    ])

test_transform = transforms.Compose([
                    transforms.ToPILImage(),
                    transforms.Resize([CFG['IMG_SIZE'], CFG['IMG_SIZE']]),
                    transforms.ToTensor(),
                    transforms.Normalize(mean=(0.5, 0.5, 0.5), std=(0.5, 0.5, 0.5))
                    ])

## Dataloader
Dataloader class는 batch기반의 딥러닝모델 학습을 위해서 mini batch를 만들어주는 역할을 합니다. 

dataloader를 통해 dataset의 전체 데이터가 batch size로 나뉘게 됩니다. 

만들었던 dataset을 input으로 넣어주면 여러 옵션(데이터 묶기, 섞기, 알아서 병렬처리)을 통해 batch를 만들어 내는 것입니다.

In [17]:
# Get Dataloader

#CustomDataset class를 통하여 train dataset생성
train_dataset = CustomDataset(train_img_path, train_label, train_mode=True, transforms=train_transform) 
#만든 train dataset를 DataLoader에 넣어 batch 만들기
train_loader = DataLoader(train_dataset, batch_size = CFG['BATCH_SIZE'], shuffle=True, num_workers=0) #BATCH_SIZE : 24

#vaildation 에서도 적용
vali_dataset = CustomDataset(vali_img_path, vali_label, train_mode=True, transforms=test_transform)
vali_loader = DataLoader(vali_dataset, batch_size = CFG['BATCH_SIZE'], shuffle=False, num_workers=0)

In [18]:
train_batches = len(train_loader)
vali_batches = len(vali_loader)

print('total train imgs :',train_len,'/ total train batches :', train_batches)
print('total valid imgs :',Vali_len, '/ total valid batches :', vali_batches)

total train imgs : 643 / total train batches : 54
total valid imgs : 214 / total valid batches : 18


배치 사이즈가 24이므로    
train은 54 묶음 vaildation은 18 묶음으로 묶인 것을 볼 수 있습니다.

## 모델 불러오기

전이학습을 위한 모델을 load 해줍니다.

Pytorch 의 models 메소드를 사용하면 손쉽게 외부의 모델을 불러올 수 있습니다.

이번 베이스라인에서는 efficientnet_b3 모델을 사용해 볼 것입니다.

사전 학습 모델을 사용하는 것은 부정행위에 해당하니, pretrained 파라미터를 False 로 설정해야 합니다!

pretrained 파라미터를 True 로 설정한다면, ImageNet 이라는 데이터셋을 대상으로 학습된 모델이 load 됩니다.

반면, pretrained 파라미터를 False 로 설정한다면, 모델의 구조만 load 되고 모델의 가중치 들은 load 되지 않습니다.

In [20]:
from torchvision import models
from torchvision.models import efficientnet_b3 as efficientnet

model = models.efficientnet_b3(pretrained=False)

In [21]:
model.classifier

Sequential(
  (0): Dropout(p=0.3, inplace=True)
  (1): Linear(in_features=1536, out_features=1000, bias=True)
)

모델에 데이터를 학습하기 위해서는 모델의 마지막 layer 의 output size 와 분류할 라벨의 수를 입력해주어야 합니다.

In [30]:
import torch.nn as nn
from torch.nn import functional as F
from torch.nn import CrossEntropyLoss
import torch.optim as optim

model.fc = nn.Linear(1000, 11)
model = model.to(device)

모델의 파라미터들을 설정해줍니다.

## 모델 학습

이제 모델 학습을 하기 위해 매개변수를 정의해보도록 하겠습니다.

In [ ]:
import torch.optim as optim # 최적화 알고리즘들이 포함힘

criterion = torch.nn.CrossEntropyLoss()
optimizer = torch.optim.Adam(params=model.parameters(), lr= CFG["LEARNING_RATE"] )#0.001
scheduler = None

train 메소드를 통하여 train을 학습 시켜 vaildation으로 평가하는 메소드를 작성해보겠습니다.

In [31]:
def train(model, optimizer, train_loader, scheduler, device): 
    model.to(device)
    n = len(train_loader)
    
    #Loss Function 정의
    #criterion = nn.CrossEntropyLoss().to(device)
    best_acc = 0
    
    for epoch in range(1,CFG["EPOCHS"]+1): #에포크 설정
        model.train() #모델 학습
        running_loss = 0.0
            
        for img, label in tqdm(iter(train_loader)):
            img, label = img.to(device), label.to(device) #배치 데이터
            optimizer.zero_grad() #배치마다 optimizer 초기화
        
            # Data -> Model -> Output
            logit = model(img) #예측값 산출
            loss = criterion(logit, label) #손실함수 계산
            
            # 역전파
            loss.backward() #손실함수 기준 역전파 
            optimizer.step() #가중치 최적화
            running_loss += loss.item()
              
        print('[%d] Train loss: %.10f' %(epoch, running_loss / len(train_loader)))
        
        if scheduler is not None:
            scheduler.step()
            
        #Validation set 평가
        model.eval() #evaluation 과정에서 사용하지 않아야 하는 layer들을 알아서 off 시키도록 하는 함수
        vali_loss = 0.0
        correct = 0
        with torch.no_grad(): #파라미터 업데이트 안하기 때문에 no_grad 사용
            for img, label in tqdm(iter(vali_loader)):
                img, label = img.to(device), label.to(device)

                logit = model(img)
                vali_loss += criterion(logit, label)
                pred = logit.argmax(dim=1, keepdim=True)  #11개의 class중 가장 값이 높은 것을 예측 label로 추출
                correct += pred.eq(label.view_as(pred)).sum().item() #예측값과 실제값이 맞으면 1 아니면 0으로 합산
        vali_acc = 100 * correct / len(vali_loader.dataset)
        print('Vail set: Loss: {:.4f}, Accuracy: {}/{} ( {:.0f}%)\n'.format(vali_loss / len(vali_loader), correct, len(vali_loader.dataset), 100 * correct / len(vali_loader.dataset)))
        
        #베스트 모델 저장
        if best_acc < vali_acc:
            best_acc = vali_acc
            torch.save(model.state_dict(), './saved/best_model.pth') #이 디렉토리에 best_model.pth을 저장
            print('Model Saved.')

classification 문제이기 때문에 평가지표로는 Accuarcy를 사용하여 모델의 정확도를 산출하였습니다.

In [32]:
train(model, optimizer, train_loader, scheduler, device)

  0%|          | 0/54 [00:00<?, ?it/s]

[1] Train loss: 2.5656345597


  0%|          | 0/18 [00:00<?, ?it/s]

Vail set: Loss: 18.3208, Accuracy: 25/215 ( 12%)

Model Saved.


  0%|          | 0/54 [00:00<?, ?it/s]

[2] Train loss: 2.4322992961


  0%|          | 0/18 [00:00<?, ?it/s]

Vail set: Loss: 2.3891, Accuracy: 29/215 ( 13%)

Model Saved.


  0%|          | 0/54 [00:00<?, ?it/s]

[3] Train loss: 2.3632486220


  0%|          | 0/18 [00:00<?, ?it/s]

Vail set: Loss: 2.6524, Accuracy: 34/215 ( 16%)

Model Saved.


  0%|          | 0/54 [00:00<?, ?it/s]

[4] Train loss: 2.1771570775


  0%|          | 0/18 [00:00<?, ?it/s]

Vail set: Loss: 2.2911, Accuracy: 33/215 ( 15%)



  0%|          | 0/54 [00:00<?, ?it/s]

[5] Train loss: 2.1643342530


  0%|          | 0/18 [00:00<?, ?it/s]

Vail set: Loss: 1.9099, Accuracy: 41/215 ( 19%)

Model Saved.


  0%|          | 0/54 [00:00<?, ?it/s]

[6] Train loss: 1.9967296190


  0%|          | 0/18 [00:00<?, ?it/s]

Vail set: Loss: 1.9408, Accuracy: 57/215 ( 27%)

Model Saved.


  0%|          | 0/54 [00:00<?, ?it/s]

[7] Train loss: 1.9540549075


  0%|          | 0/18 [00:00<?, ?it/s]

Vail set: Loss: 1.9223, Accuracy: 39/215 ( 18%)



  0%|          | 0/54 [00:00<?, ?it/s]

[8] Train loss: 1.8215324945


  0%|          | 0/18 [00:00<?, ?it/s]

Vail set: Loss: 1.7438, Accuracy: 58/215 ( 27%)

Model Saved.


  0%|          | 0/54 [00:00<?, ?it/s]

[9] Train loss: 1.6996934833


  0%|          | 0/18 [00:00<?, ?it/s]

Vail set: Loss: 1.7026, Accuracy: 55/215 ( 26%)



  0%|          | 0/54 [00:00<?, ?it/s]

[10] Train loss: 1.6081311195


  0%|          | 0/18 [00:00<?, ?it/s]

Vail set: Loss: 1.6803, Accuracy: 74/215 ( 34%)

Model Saved.


  0%|          | 0/54 [00:00<?, ?it/s]

[11] Train loss: 1.5765591937


  0%|          | 0/18 [00:00<?, ?it/s]

Vail set: Loss: 1.3190, Accuracy: 91/215 ( 42%)

Model Saved.


  0%|          | 0/54 [00:00<?, ?it/s]

[12] Train loss: 1.4788019867


  0%|          | 0/18 [00:00<?, ?it/s]

Vail set: Loss: 1.4486, Accuracy: 90/215 ( 42%)



  0%|          | 0/54 [00:00<?, ?it/s]

[13] Train loss: 1.3816111573


  0%|          | 0/18 [00:00<?, ?it/s]

Vail set: Loss: 1.3512, Accuracy: 87/215 ( 40%)



  0%|          | 0/54 [00:00<?, ?it/s]

[14] Train loss: 1.2490052294


  0%|          | 0/18 [00:00<?, ?it/s]

Vail set: Loss: 1.2024, Accuracy: 111/215 ( 52%)

Model Saved.


  0%|          | 0/54 [00:00<?, ?it/s]

[15] Train loss: 1.2514997632


  0%|          | 0/18 [00:00<?, ?it/s]

Vail set: Loss: 1.5822, Accuracy: 93/215 ( 43%)



  0%|          | 0/54 [00:00<?, ?it/s]

[16] Train loss: 1.1561281184


  0%|          | 0/18 [00:00<?, ?it/s]

Vail set: Loss: 1.2881, Accuracy: 109/215 ( 51%)



  0%|          | 0/54 [00:00<?, ?it/s]

[17] Train loss: 1.1824170506


  0%|          | 0/18 [00:00<?, ?it/s]

Vail set: Loss: 1.2877, Accuracy: 109/215 ( 51%)



  0%|          | 0/54 [00:00<?, ?it/s]

[18] Train loss: 1.0781494511


  0%|          | 0/18 [00:00<?, ?it/s]

Vail set: Loss: 1.2547, Accuracy: 104/215 ( 48%)



  0%|          | 0/54 [00:00<?, ?it/s]

[19] Train loss: 0.9934439598


  0%|          | 0/18 [00:00<?, ?it/s]

Vail set: Loss: 1.1814, Accuracy: 118/215 ( 55%)

Model Saved.


  0%|          | 0/54 [00:00<?, ?it/s]

[20] Train loss: 0.7660036678


  0%|          | 0/18 [00:00<?, ?it/s]

Vail set: Loss: 1.2443, Accuracy: 124/215 ( 58%)

Model Saved.


  0%|          | 0/54 [00:00<?, ?it/s]

[21] Train loss: 0.9518345705


  0%|          | 0/18 [00:00<?, ?it/s]

Vail set: Loss: 1.2795, Accuracy: 119/215 ( 55%)



  0%|          | 0/54 [00:00<?, ?it/s]

[22] Train loss: 0.8807636927


  0%|          | 0/18 [00:00<?, ?it/s]

Vail set: Loss: 1.0672, Accuracy: 129/215 ( 60%)

Model Saved.


  0%|          | 0/54 [00:00<?, ?it/s]

[23] Train loss: 0.6947549057


  0%|          | 0/18 [00:00<?, ?it/s]

Vail set: Loss: 1.0825, Accuracy: 130/215 ( 60%)

Model Saved.


  0%|          | 0/54 [00:00<?, ?it/s]

[24] Train loss: 0.7048165699


  0%|          | 0/18 [00:00<?, ?it/s]

Vail set: Loss: 1.1949, Accuracy: 136/215 ( 63%)

Model Saved.


  0%|          | 0/54 [00:00<?, ?it/s]

[25] Train loss: 0.6005065030


  0%|          | 0/18 [00:00<?, ?it/s]

Vail set: Loss: 1.0590, Accuracy: 141/215 ( 66%)

Model Saved.


  0%|          | 0/54 [00:00<?, ?it/s]

[26] Train loss: 0.6234038660


  0%|          | 0/18 [00:00<?, ?it/s]

Vail set: Loss: 1.1091, Accuracy: 137/215 ( 64%)



  0%|          | 0/54 [00:00<?, ?it/s]

[27] Train loss: 0.6412223876


  0%|          | 0/18 [00:00<?, ?it/s]

Vail set: Loss: 1.0061, Accuracy: 148/215 ( 69%)

Model Saved.


  0%|          | 0/54 [00:00<?, ?it/s]

[28] Train loss: 0.5437638263


  0%|          | 0/18 [00:00<?, ?it/s]

Vail set: Loss: 0.9072, Accuracy: 158/215 ( 73%)

Model Saved.


  0%|          | 0/54 [00:00<?, ?it/s]

[29] Train loss: 0.4045370056


  0%|          | 0/18 [00:00<?, ?it/s]

Vail set: Loss: 1.0197, Accuracy: 148/215 ( 69%)



  0%|          | 0/54 [00:00<?, ?it/s]

[30] Train loss: 0.4756949040


  0%|          | 0/18 [00:00<?, ?it/s]

Vail set: Loss: 1.5881, Accuracy: 114/215 ( 53%)



  0%|          | 0/54 [00:00<?, ?it/s]

[31] Train loss: 0.6310135637


  0%|          | 0/18 [00:00<?, ?it/s]

Vail set: Loss: 1.2455, Accuracy: 135/215 ( 63%)



  0%|          | 0/54 [00:00<?, ?it/s]

[32] Train loss: 0.5273576881


  0%|          | 0/18 [00:00<?, ?it/s]

Vail set: Loss: 1.0516, Accuracy: 141/215 ( 66%)



  0%|          | 0/54 [00:00<?, ?it/s]

[33] Train loss: 0.4163163681


  0%|          | 0/18 [00:00<?, ?it/s]

Vail set: Loss: 1.0260, Accuracy: 153/215 ( 71%)



  0%|          | 0/54 [00:00<?, ?it/s]

[34] Train loss: 0.3775716007


  0%|          | 0/18 [00:00<?, ?it/s]

Vail set: Loss: 1.0768, Accuracy: 145/215 ( 67%)



  0%|          | 0/54 [00:00<?, ?it/s]

[35] Train loss: 0.3731325339


  0%|          | 0/18 [00:00<?, ?it/s]

Vail set: Loss: 1.2807, Accuracy: 139/215 ( 65%)



  0%|          | 0/54 [00:00<?, ?it/s]

[36] Train loss: 0.3889204191


  0%|          | 0/18 [00:00<?, ?it/s]

Vail set: Loss: 1.1691, Accuracy: 150/215 ( 70%)



  0%|          | 0/54 [00:00<?, ?it/s]

[37] Train loss: 0.4508861930


  0%|          | 0/18 [00:00<?, ?it/s]

Vail set: Loss: 1.1352, Accuracy: 147/215 ( 68%)



  0%|          | 0/54 [00:00<?, ?it/s]

[38] Train loss: 0.3592968580


  0%|          | 0/18 [00:00<?, ?it/s]

Vail set: Loss: 1.0816, Accuracy: 144/215 ( 67%)



  0%|          | 0/54 [00:00<?, ?it/s]

[39] Train loss: 0.2248456906


  0%|          | 0/18 [00:00<?, ?it/s]

Vail set: Loss: 1.1456, Accuracy: 157/215 ( 73%)



  0%|          | 0/54 [00:00<?, ?it/s]

[40] Train loss: 0.3709733281


  0%|          | 0/18 [00:00<?, ?it/s]

Vail set: Loss: 1.1131, Accuracy: 145/215 ( 67%)



  0%|          | 0/54 [00:00<?, ?it/s]

[41] Train loss: 0.2387158730


  0%|          | 0/18 [00:00<?, ?it/s]

Vail set: Loss: 1.1725, Accuracy: 143/215 ( 67%)



  0%|          | 0/54 [00:00<?, ?it/s]

[42] Train loss: 0.1930999178


  0%|          | 0/18 [00:00<?, ?it/s]

Vail set: Loss: 1.1664, Accuracy: 152/215 ( 71%)



  0%|          | 0/54 [00:00<?, ?it/s]

[43] Train loss: 0.1610608929


  0%|          | 0/18 [00:00<?, ?it/s]

Vail set: Loss: 1.0805, Accuracy: 148/215 ( 69%)



  0%|          | 0/54 [00:00<?, ?it/s]

[44] Train loss: 0.2892991534


  0%|          | 0/18 [00:00<?, ?it/s]

Vail set: Loss: 1.1846, Accuracy: 152/215 ( 71%)



  0%|          | 0/54 [00:00<?, ?it/s]

[45] Train loss: 0.2680082468


  0%|          | 0/18 [00:00<?, ?it/s]

Vail set: Loss: 1.0472, Accuracy: 157/215 ( 73%)



  0%|          | 0/54 [00:00<?, ?it/s]

[46] Train loss: 0.1715788394


  0%|          | 0/18 [00:00<?, ?it/s]

Vail set: Loss: 0.8986, Accuracy: 162/215 ( 75%)

Model Saved.


  0%|          | 0/54 [00:00<?, ?it/s]

[47] Train loss: 0.2311460532


  0%|          | 0/18 [00:00<?, ?it/s]

Vail set: Loss: 1.3049, Accuracy: 147/215 ( 68%)



  0%|          | 0/54 [00:00<?, ?it/s]

[48] Train loss: 0.2140235253


  0%|          | 0/18 [00:00<?, ?it/s]

Vail set: Loss: 1.7686, Accuracy: 144/215 ( 67%)



  0%|          | 0/54 [00:00<?, ?it/s]

[49] Train loss: 0.2012073077


  0%|          | 0/18 [00:00<?, ?it/s]

Vail set: Loss: 1.2530, Accuracy: 147/215 ( 68%)



  0%|          | 0/54 [00:00<?, ?it/s]

[50] Train loss: 0.2123492892


  0%|          | 0/18 [00:00<?, ?it/s]

Vail set: Loss: 1.4316, Accuracy: 137/215 ( 64%)



에포크가 46일때 Vaildation Accuracy가 75%로 best_model에 선정되어 저장되었습니다.

## 추론하기

이제 학습된 best_model을 가지고 test 셋의 라벨을 추론해보도록 하겠습니다.

In [34]:
def predict(model, test_loader, device):
    model.eval()
    model_pred = []
    with torch.no_grad():
        for img in tqdm(iter(test_loader)):
            img = img.to(device)

            pred_logit = model(img)
            pred_logit = pred_logit.argmax(dim=1, keepdim=True).squeeze(1)

            model_pred.extend(pred_logit.tolist())
    return model_pred

In [38]:
test_dataset = CustomDataset(test_img_path, None, train_mode=False, transforms=test_transform)
test_loader = DataLoader(test_dataset, batch_size = CFG['BATCH_SIZE'], shuffle=False, num_workers=0)

# Validation Accuracy가 가장 뛰어난 모델을 불러옵니다.
checkpoint = torch.load('./saved/best_model.pth')
model = models.efficientnet_b3(pretrained=False)
model.fc = nn.Linear(1000, 11)
model = model.to(device)
model.load_state_dict(checkpoint)

# Inference
preds = predict(model, test_loader, device)
preds[0:5]

  0%|          | 0/18 [00:00<?, ?it/s]

[1, 10, 1, 6, 9]

값이 배열안에 정상적으로 잘 들어간 것을 확인할 수 있습니다.

## 제출하기

submission에 예측한 값 preds를 넣어줍시다

In [39]:
submission = pd.read_csv('data/sample_submission.csv')
submission['label'] = preds

이제 제출을 위해 라벨을 다시 복원 시켜 줍니다.

앞서 10-1을 10으로, 10-2를 0으로 바꿔주었던 값을 다시 원래의 값으로 바꿔주겠습니다.

또한 label 열의 타입을 int에서 object로 수정해 주겠습니다.

In [40]:
submission['label'][submission['label'] == 10] = '10-1' ## label : 10 -> '10-1'
submission['label'][submission['label'] == 0] = '10-2' ## Label : 0 -> '10-2'
submission['label'] = submission['label'].apply(lambda x : str(x)) ## Dtype : int -> object

/tmp/ipykernel_6419/3411836075.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  submission['label'][submission['label'] == 10] = '10-1' ## label : 10 -> '10-1'


In [41]:
submission.head()

file_name label
0   001.png     1
1   002.png  10-1
2   003.png     1
3   004.png     6
4   005.png     9

submission을 submit2란 이름으로 csv 파일로 저장합니다.   

index=False란 추가적인 id를 부여할 필요가 없다는 뜻입니다.   

정확한 채점을 위해 꼭 index=False를 넣어주세요.

In [42]:
submission.to_csv('submit2.csv', index=False)

이렇게 생성된 submit2.csv 파일을 데이콘 대회 페이지에 업로드 & 제출하여 결과를 확인해보세요!

문제를 해결하기 위한 여러분의 방법을 코드 공유 게시판에 공유해주세요

좋아요와 댓글을 합산하여 가장 높은 점수를 얻으신 분께 데이콘 후드가 제공됩니다!